w3-compute.md

```diff
@@ -0,0 +1,125 @@
+# W3 Compute Protocol
+
+![status:draft](https://img.shields.io/badge/status-reliable-green.svg?style=flat-square)
+
+## Editors
+
+- [Vasco Santos], [Protocol Labs]
+
+## Authors
+
+- [Vasco Santos], [Protocol Labs]
+
+# Abstract
+
+This spec describes a [UCAN] protocol allowing an implementer to perform simple computations over data on behalf of an issuer.
+
+## Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+# Table of Contents
+
+- [Introduction](#introduction)
+- [Capabilities](#capabilities)
+  - [`compute/` namespace](#compute-namespace)
+    - [`compute/*`](#compute)
+    - [`compute/piececid`](#computepiececid)
+- [Schema](#schema)
+  - [`compute/piececid` schema](#computepiececid-schema)
+
+# Introduction
+
+Within the w3up protocol flows, some computations MUST be performed over data. These case range from computing proofs to data indexes, as well as to verify client side offered computations.
+
+The `w3-compute` protocol aims to enable clients to hire compute services to delegate some work, as well as for `w3-up` platform to hire third party compute services to verify client side offered computations if desirable.
+
+Note that the discovery process by actors looking for services providing given computations is for now out of scope of this spec.
+
+# Capabilities
+
+## `compute/` namespace
+
+The `compute/` namespace contains capabilities relating to computations.
+
+## `compute/*`
+
+> Delegate all capabilities in the `compute/` namespace
+
+The `compute/*` capability is the "top" capability of the `compute/*` namespace. `compute/*` can be delegated to a user agent, but cannot be invoked directly. Instead, it allows the agent to derive any capability in the `compute/` namespace, provided the resource URI matches the one in the `compute/*` capability delegation.
+
+In other words, if an agent has a delegation for `compute/*` for a given space URI, they can invoke any capability in the `compute/` namespace using that space as the resource.
+
+## `compute/piececid`
+
+Request computation of a PieceCIDv2 per [FRC-0069](https://github.com/filecoin-project/FIPs/blob/master/FRCs/frc-0069.md). A CID representation for the FR32 padded sha256-trunc254-padded binary merkle trees used in Filecoin Piece Commitments.
+
+> `did:key:zAliceAgent` invokes `compute/piececid` capability provided by `did:web:web3.storage`
+
+```json
+{
+  "iss": "did:key:zAliceAgent",
+  "aud": "did:web:web3.storage",
+  "att": [
+    {
+      "with": "did:key:zAliceAgent",
+      "can": "compute/piececid",
+      "nb": {
+        /* CID of the uploaded content */
+        "content": { "/": "bag...car" }
```

Comment:
I think this should be a link to a UCAN that gives you read access to the content. While we do allow public reads right now, this seems like an overkill. However if reads are charged and potentially accelerated it would make a lot of sense to pass "readable resource" itself which in UCAN word would be a capability on a resource giving you a read access.

Reply:
I often like to think of capability groups like TS interfaces with some methods e.g. upload/* is something like

```ts
interface Upload {
   add(upload): Promise<Result<...>>
   list(): Promise<....>
   ....
}
```

When I delegate access I give you reference to either `Upload` instance or just selected methods from it.

Reply:
I think here is where we will need to map lot's of assumptions around what are the requirements and needs. I have seen this as a service that can choose where to read from according to their preferences. If I run in CF I prefer to read from R2, but the service is deployed anywhere. So, would not really be the caller to infer that.

Probably we need to consider better what we want and what are the requirements from perf and cost before deciding how these flows would be